components/contribution-flow/index.js

```diff
@@ -421,7 +421,18 @@ class CreateOrderPage extends React.Component {
         this.handleSuccess(orderCreated);
       }
     } catch (e) {
-      this.setState({ submitting: false, error: e.message });
+      this.setState(state => ({
+        ...state,
+        submitting: false,
+        error: e.message,
+        stepPayment: {
+          ...state.stepPayment,
+          paymentMethod: {
+            ...state.stepPayment.paymentMethod,
+            token: null,
+          },
+        },
```

Comment:
This might work but this doesn't look good. Can you try a simpler code? 

Don't try to do everything in a single statement and try with `return state` at the end.

Reply:
@znarf 
is this going to work, or is this so similar to the previous solution ? 
I think it is much more readable now but is it enough, can it be improved more ? 
```
this.setState(state => {
        const stepPayment = {
          ...state.stepPayment, 
          paymentMethod: {...state.stepPayment.paymentMethod}
        };
        stepPayment.paymentMethod.token = null;
        return {...state, submitting: false, error: e.message, stepPayment };
});
```

Reply:
Seems we don't need to pass the full state, so we can remove `...state`.

> The output of the updater is shallowly merged with state. https://reactjs.org/docs/react-component.html#setstate

Then, I was concerned before that you were using too many destructuring assignments, but now that I had a closer look, I understand better why they are necessary here.

```
this.setState(state => {
  const stepPayment = {
    ...state.stepPayment, 
    paymentMethod: {
      ...state.stepPayment.paymentMethod,
      token: null
    }
  };
  return { submitting: false, error: e.message, stepPayment };
});
```

Then it's effectively not far from your original version. Would be interesting to have the feedback from @Betree also.

Reply:
`...state` is indeed not necessary (I learned that today).

I'm ok with this code, an alternative could be:


```es6
const stepPayment = cloneDeep(state.stepPayment)
set(stepPayment, 'paymentMethod.token', null)
this.setState(state => ({
  (...),
  stepPayment,
})
```

Reply:
Ok, let's move forward with one of these two options.

```
this.setState(state => {
  const stepPayment = {
    ...state.stepPayment, 
    paymentMethod: {
      ...state.stepPayment.paymentMethod,
      token: null
    }
  };
  return { submitting: false, error: e.message, stepPayment };
});
```

or:

```
      this.setState(state => ({
        submitting: false,
        error: e.message,
        stepPayment: {
          ...state.stepPayment,
          paymentMethod: {
            ...state.stepPayment.paymentMethod,
            token: null,
          },
        },
```

Reply:
ok then, let me make these changes and it'll be ready in this evening. 
@Betree 's solution is perfect but I think this can be read easier by people and more traditional way. But it's all up to you.

```
this.setState(state => {
  const stepPayment = {
    ...state.stepPayment, 
    paymentMethod: {
      ...state.stepPayment.paymentMethod,
      token: null
    }
  };
  return { submitting: false, error: e.message, stepPayment };
});
```